### PR TITLE
Tracking doc: rewrite `@wdio/applitools-service` package into TypeScript

### LINKED ISSUE #5827: Tracking doc: rewrite `@wdio/applitools-service` package into TypeScript
This is a tracking issue for organising the effort of re-writing the code base from vanilla JS into TypeScript. If you are reading this and are interested getting involved, this is a great chance to do so. Here are the rules:

- if you are interested converting source files from `/packages/wdio-applitools-service` into TypeScript please comment on this thread and mention the files you want to convert to avoid duplicate work
- there is no reason to convert the whole package at once - smaller PRs make it easier to review the code
- if you make a PR, reference this issue so we can keep track of the progress
- if you make a PR, label it with `TypeScript` and the package name `wdio-applitools-service` so we can easier find all PRs
- if you make a PR, assign it to the `TypeScript Rewrite` project

Happy rewriting 😅

### PR DESCRIPTION
**Pull request by mciastek**
## Proposed changes

[//]: # (Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.)

Rewrite of `@wdio/applitools-service` package into TypeScript. Resolves #5827

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/master/CONTRIBUTING.md) doc
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] I have added proper type definitions for new commands (if appropriate)

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### CODE REVIEW

File: packages/wdio-applitools-service/applitools-service.d.ts [hunk 1]
```diff
@@ -19,17 +19,17 @@ interface ApplitoolsConfig {
...
     };
     /**
      * Use proxy for http/https connections with Applitools.
      */
     proxy?: {
-        url: string;
-        username?: string;
-        password?: string;
-        isHttpOnly?: boolean;
+        url: string | boolean;
```
Comment (christian-bromann) [inline packages/wdio-applitools-service/applitools-service.d.ts:30]: How can this be a boolean?

File: packages/wdio-applitools-service/src/index.ts [hunk 2]
```diff
@@ -72,8 +108,8 @@ export default class ApplitoolsService {
         })
     }
 
-    beforeTest(test) {
-        if (!this.isConfigured) {
+    beforeTest(test?: { title: string, parent: string }) {
```
Comment (christian-bromann) [inline packages/wdio-applitools-service/src/index.ts:88]: The parameter will always be provided.

**Review by christian-bromann — CHANGES_REQUESTED** [review]
Some minor comments

File: packages/wdio-applitools-service/applitools-service.d.ts (hunk 1 above)
Comment (mciastek) [reply packages/wdio-applitools-service/applitools-service.d.ts:30]: I was trying to follow documentation and source code of applitools at the same time. Honestly, it's a mess 😅  There are several implementations of `setProxy()` method that takes either `ProxySettings` object or a number of arguments. I will this one, thanks!

File: packages/wdio-applitools-service/src/index.ts (hunk 2 above)
Comment (mciastek) [reply packages/wdio-applitools-service/src/index.ts:88]: There is a test that calls this method without any arguments - https://github.com/webdriverio/webdriverio/blob/cc3c86dad3cd2b2636eb45c578ee1e8ac760c100/packages/wdio-applitools-service/tests/service.test.ts#L208

Is it correct?

File: packages/wdio-applitools-service/tests/service.test.ts [hunk 3]
```diff
@@ -1,6 +1,12 @@
 import ApplitoolsService from '../src'
 
+const globalAny: any = global
+
 class BrowserMock {
+    [key: string]: any;
+    addCommand: any;
```
Comment (mgrybyk) [inline packages/wdio-applitools-service/tests/service.test.ts:7]: these ones are not needed if you have `[key: string]: any;`

File: packages/wdio-applitools-service/src/@types/@applitools/eyes-webdriverio.d.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+interface Eyes {
+    new(serverUrl: string, isDisabled: boolean): Eyes;
+
+    setServerUrl(value: string): void;
+    setApiKey(value: string): void;
+    setProxy(value: ProxyOptions): void;
+
+    open(driver: any, appName: string, testName: string, viewportSize?: { width: number, height: number }): Promise<any>;
```
Comment (mgrybyk) [inline packages/wdio-applitools-service/src/@types/@applitools/eyes-webdriverio.d.ts:22]: is it possible to replace `any` with specific types?

File: packages/wdio-applitools-service/applitools-service.d.ts [hunk 5]
```diff
@@ -19,17 +20,17 @@ interface ApplitoolsConfig {
...
      * Use proxy for http/https connections with Applitools.
      */
     proxy?: {
         url: string;
-        username?: string;
-        password?: string;
-        isHttpOnly?: boolean;
+        username: string;
+        password: string;
+        isHttpOnly: boolean;
```
Comment (mgrybyk) [inline packages/wdio-applitools-service/applitools-service.d.ts:33]: should all of them be really required?

File: packages/wdio-applitools-service/applitools-service.d.ts [hunk 6]
```diff
@@ -3,6 +3,7 @@
 declare module WebdriverIO {
     interface ServiceOption extends ApplitoolsConfig {}
     interface Browser extends ApplitoolsBrowser {}
+    interface Element {}
```
Comment (mgrybyk) [inline packages/wdio-applitools-service/applitools-service.d.ts:6]: why do you need this?

File: packages/wdio-applitools-service/src/@types/@applitools/eyes-webdriverio.d.ts [hunk 7]
```diff
@@ -0,0 +1,31 @@
+/// <reference types="@applitools/visual-grid-client"/>
+
+declare module '@applitools/eyes-webdriverio';
+
+type Region = Eyes.Check.Region | WebdriverIO.Element | string;
+type Frame = WebdriverIO.Element | string;
+
+interface ProxyOptions {
```
Comment (mgrybyk) [inline packages/wdio-applitools-service/src/@types/@applitools/eyes-webdriverio.d.ts:8]: Can you reuse proxy from `ApplitoolsConfig`? Maybe you can define it there as interface

**Review by mgrybyk — CHANGES_REQUESTED** [review]
Please fix build errors!
```
Error: packages/wdio-applitools-service/src/index.ts(1,20): error TS2307: Cannot find module '@wdio/logger' or its corresponding type declarations.
Error: packages/wdio-applitools-service/src/index.ts(20,14): error TS2304: Cannot find name 'ApplitoolsConfig'.
Error: packages/wdio-applitools-service/src/index.ts(22,24): error TS2304: Cannot find name 'ApplitoolsConfig'.
Error: packages/wdio-applitools-service/src/index.ts(25,26): error TS2304: Cannot find name 'ApplitoolsConfig'.
```
https://github.com/webdriverio/webdriverio/pull/5981/checks?check_run_id=1246338459

**Comment by mciastek** [comment]
> Please fix build errors!
> 
> ```
> Error: packages/wdio-applitools-service/src/index.ts(1,20): error TS2307: Cannot find module '@wdio/logger' or its corresponding type declarations.
> Error: packages/wdio-applitools-service/src/index.ts(20,14): error TS2304: Cannot find name 'ApplitoolsConfig'.
> Error: packages/wdio-applitools-service/src/index.ts(22,24): error TS2304: Cannot find name 'ApplitoolsConfig'.
> Error: packages/wdio-applitools-service/src/index.ts(25,26): error TS2304: Cannot find name 'ApplitoolsConfig'.
> ```
> 
> https://github.com/webdriverio/webdriverio/pull/5981/checks?check_run_id=1246338459

I can't see mentioned errors. Could you tell me what build failed?

**Comment by mgrybyk** [comment]
Just run npm run build. 
You'll see the errors, the build errors are ignored now so it won't fail (will be fixed soon).
You can also cd to the package folder and run tsc

File: packages/wdio-applitools-service/applitools-service.d.ts (hunk 6 above)
Comment (mciastek) [reply packages/wdio-applitools-service/applitools-service.d.ts:6]: Because here we define a new `WebdriverIO` namespace, that has new `Browser` and `ServiceOption`, and no `Element` defined. Then `WebdriverIO.Element` throws an error. I can try to find a better way.

File: packages/wdio-applitools-service/applitools-service.d.ts (hunk 5 above)
Comment (mciastek) [reply packages/wdio-applitools-service/applitools-service.d.ts:33]: Here's a definition of ProxySettings in applitools - https://github.com/applitools/eyes.sdk.javascript1/blob/master/packages/eyes-sdk-core/lib/config/Configuration.js#L346-L362. I'm not sure if it's a correct version of SDK webdriverio uses, but it looks like, when an object is passed, all properties are required.

**Comment by mciastek** [comment]
> Just run npm run build.
> You'll see the errors, the build errors are ignored now so it won't fail (will be fixed soon).
> You can also cd to the package folder and run tsc

I managed to solve a couple of issues. There are others, like `Cannot find module '@wdio/logger' or its corresponding type declarations.`, to be resolved.

**Comment by mgrybyk** [comment]
Make sure to rebase with master and run setup script

**Comment by mciastek** [comment]
> Make sure to rebase with master and run setup script

I rebased onto the current master and it seems that `@wdio/logger` is not visible for TypeScript in `wdio-applitools-service`. I'm not accustomed to type structure in webdriverio, so I'm not sure where is the problem. I tried to follow `wdio-shared-store-service` file structure and still I get this missing types problem. @mgrybyk I noticed that you did [this trick](https://github.com/webdriverio/webdriverio/pull/5927/files#diff-ade510e66107f6f7ae4bbaea000a1dc179f3f54a28eced06493eb403fab6cb56) to solve this issue, but this file has been deleted in [other PR](https://github.com/webdriverio/webdriverio/pull/5853/files#diff-ade510e66107f6f7ae4bbaea000a1dc179f3f54a28eced06493eb403fab6cb56), so maybe there's other way to solve this.

File: packages/wdio-applitools-service/src/index.ts (hunk 2 above)
Comment (christian-bromann) [reply packages/wdio-applitools-service/src/index.ts:88]: we should adjust the test to pass in a data payload that fits our expectations.

**Comment by christian-bromann** [comment]
@mciastek can you check how other packages have done it? The Applitools service shouldn't be much different to those packages. If `@wdio/logger` is a dependency of the applitools service you should get its types too. Maybe try to rerun `npm run setup-full`

**Comment by mciastek** [comment]
@christian-bromann @mgrybyk PR is ready for another review.

File: packages/wdio-applitools-service/src/@types/@applitools/eyes-webdriverio.d.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
+/// <reference types="@applitools/visual-grid-client"/>
+/// <reference types="../applitools-service"/>
```
Comment (christian-bromann) [inline packages/wdio-applitools-service/src/@types/@applitools/eyes-webdriverio.d.ts:2]: Are these needed?

File: packages/wdio-applitools-service/src/@types/applitools-service.d.ts [hunk 9]
```diff
@@ -0,0 +1 @@
+../../applitools-service.d.ts
```
Comment (christian-bromann) [inline packages/wdio-applitools-service/src/@types/applitools-service.d.ts:1]: Why is this needed?

File: scripts/build.js [hunk 10]
```diff
@@ -14,16 +14,53 @@ if (HAS_WATCH_FLAG) {
...
+        const tokens = pkg.split('-')
+
+        if (tokens.length > 2) {
+            acc[1].push(pkg)
+        } else {
+            acc[0].push(pkg)
+        }
+
+        return acc
+    }, [[], []])
```
Comment (christian-bromann) [inline scripts/build.js:57]:
```suggestion
    .reduce((acc, pkg) => {
        const tokens = pkg.split('-')
        acc[tokens.length > 2 ? 1 : 0].push(pkg)
        return acc
    }, [[], []])
```

File: packages/wdio-applitools-service/src/@types/applitools-service.d.ts (hunk 9 above)
Comment (mciastek) [reply packages/wdio-applitools-service/src/@types/applitools-service.d.ts:1]: We need to have package's typing accessible in `src` directory and outside of it. I guess it's a way to generate typings for non-TS files, but I'm not sure if it's required for rewritten files.

File: packages/wdio-applitools-service/src/@types/@applitools/eyes-webdriverio.d.ts (hunk 8 above)
Comment (mciastek) [reply packages/wdio-applitools-service/src/@types/@applitools/eyes-webdriverio.d.ts:2]: I'd like to avoid duplicating `ProxySettings` - https://github.com/webdriverio/webdriverio/pull/5981/files#diff-b360aff2f1d276bb6ea28b5b1ab3888db08d0a2be1da0dd51405e0bf26e34987R8-R14. Both `eyes-webdriverio.d.ts` and `applitools-service.d.ts` uses `ProxySettings` interface.

**Review by christian-bromann — APPROVED** [review]
Thank you so much 👍
